core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -2467,6 +2506,367 @@ public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorCo
+                }
+
+                if (parameter instanceof LambdaExpression || parameter instanceof BindExpression) { // TODO do we need this limitation?
+                    throw semanticException(NOT_SUPPORTED, parameter, "%s is not supported as JSON path parameter", parameter.getClass().getSimpleName());
+                }
+                // if the input expression is a JSON-returning function, there should be an explicit or implicit input format (spec p.817)
+                // JSON-returning functions are: JSON_OBJECT, JSON_OBJECTAGG, JSON_ARRAY, JSON_ARRAYAGG and JSON_QUERY
+                if (parameter instanceof JsonQuery && // TODO add JSON_OBJECT, JSON_OBJECTAGG, JSON_ARRAY, JSON_ARRAYAGG when supported
+                        parameterFormat.isEmpty()) {
+                    parameterFormat = Optional.of(JsonFormat.JSON);
+                }
+
+                Type parameterType = process(parameter, context);
+                if (parameterFormat.isPresent()) {
+                    // resolve function to read the parameter as JSON
+                    ResolvedFunction parameterInputFunction = getInputFunction(parameterType, parameterFormat.get(), parameter);
+                    Type expectedParameterType = parameterInputFunction.getSignature().getArgumentType(0);
+                    coerceType(parameter, parameterType, expectedParameterType, format("%s function JSON path parameter", functionName));
+                    inputFunctions.put(NodeRef.of(parameter), parameterInputFunction);
+                    passedType = JSON_2016;
+                }
+                else {
+                    if (!isStringType(parameterType) &&
+                            !isNumericType(parameterType) &&
+                            !isDateTimeType(parameterType) &&
+                            !parameterType.equals(BOOLEAN) &&
+                            !typeCoercion.canCoerce(parameterType, VARCHAR) ||
+                            parameterType.equals(INTERVAL_DAY_TIME) ||
+                            parameterType.equals(INTERVAL_YEAR_MONTH)) {
+                        throw semanticException(INVALID_FUNCTION_ARGUMENT, parameter, "Invalid type of JSON path parameter: %s", parameterType.getDisplayName());
+                    }
+                    if (isStringType(parameterType) && !isCharacterStringType(parameterType)) {
+                        throw semanticException(NOT_SUPPORTED, parameter, "Unsupported type of JSON path parameter: %s", parameterType.getDisplayName());
+                    }
+                    if (!isStringType(parameterType) &&
+                            !isNumericType(parameterType) &&
+                            !isDateTimeType(parameterType) &&
+                            !parameterType.equals(BOOLEAN)) {
+                        coerceType(parameter, parameterType, VARCHAR, "JSON path parameter");
```

Comment:
I would find logic here more readable if you used positive checks.
- if this input type ... then this passed type
- if this input type ... then this passed type.
- ...
- else throw.